### PR TITLE
Fixed #6785 -- Made QuerySet.get() fetch a limited number of rows.

### PR DESCRIPTION
**Pull request by amir-hadi**
Fixes https://code.djangoproject.com/ticket/6785 by using the hints from Shai Berger and Anssi. Needs tests with an Oracle DB please.

### CODE REVIEW

**Comment by apollo13** [comment]
buildbot, test on oracle.

**Comment by shaib** [comment]
buildbot, test on oracle.

File: django/db/models/query.py [hunk 1]
```diff
@@ -28,6 +28,10 @@
 from django.utils.functional import cached_property, partition
 from django.utils.version import get_version
 
+# The maximum number (one less than the max to be precise) of results to fetch
+# in a get() query
+MAX_GET_RESULTS = 20
```
Comment (felixxm) [inline django/db/models/query.py]: I would use `MAX_GET_RESULTS = 21`, after that code and comments can be simplified:
 - `The maximum number (one less than the max to be precise) of results to fetch` -> `The maximum number of results to fetch`,
 - `clone.query.set_limits(high=MAX_GET_RESULTS + 1)` -> `clone.query.set_limits(high=MAX_GET_RESULTS)`,
 - `num if num <= MAX_GET_RESULTS else ...` -> `num if num < MAX_GET_RESULTS else ...`, etc.

File: tests/basic/tests.py [hunk 2]
```diff
@@ -370,6 +370,28 @@ def test_delete_and_access_field(self):
         # Fields that weren't deleted aren't reloaded.
         self.assertEqual(article.pub_date, new_pub_date)
 
+    def test_multiple_objects_max_num_fetched(self):
+        """
+        #6785 - get() should fetch a limited number of results.
+        """
```
Comment (felixxm) [inline tests/basic/tests.py]: I don't see any value in this docstring please remove it.

File: django/db/models/query.py [hunk 3]
```diff
@@ -28,6 +28,10 @@
 from django.utils.functional import cached_property, partition
 from django.utils.version import get_version
 
+# The maximum number (one less than the max to be precise) of results to fetch
+# in a get() query
```
Comment (felixxm) [inline django/db/models/query.py]: Trailing dot is missing.

File: django/db/models/query.py [hunk 4]
```diff
@@ -407,8 +412,10 @@ def get(self, *args, **kwargs):
                 self.model._meta.object_name
             )
         raise self.model.MultipleObjectsReturned(
-            "get() returned more than one %s -- it returned %s!" %
-            (self.model._meta.object_name, num)
+            "get() returned more than one %s -- it returned %s!" % (
```
Comment (felixxm) [inline django/db/models/query.py]: Please use single quotes in added or changed strings.

File: django/db/models/query.py [hunk 5]
```diff
@@ -407,8 +412,10 @@ def get(self, *args, **kwargs):
...
+                self.model._meta.object_name,
+                num if num <= MAX_GET_RESULTS else 'more than %s' % MAX_GET_RESULTS
```
Comment (felixxm) [inline django/db/models/query.py]: Please add trailing comma.

**Comment by apollo13** [comment]
Why is MAX_GET_RESULTS set to 20? I do not understand that from the code nor the ticket.

**Comment by felixxm** [comment]
@apollo13 20 has been proposed by Anssi (see [discussion](https://groups.google.com/d/msg/django-developers/PkzS9Wv6hIU/ZElSGP1__JIJ)).

**Comment by apollo13** [comment]
Ah, he says:
> If somebody feels strongly about this, then we could limit the amount of fetched objects
to some smallish amount (21 for example). 

but also:
> As for .get() - I don't find the number of duplicates in the error message that useful. If you got more than one object you failed. How badly you failed isn't interesting. I can't think of any case where the  mount of objects you got actually matters.

So I'd go for two unless someone can propose a really good reason :)

**Comment by felixxm** [comment]
There is also a comment from Marc:
> I'd be inclined to agree with Anssi that we could do something like LIMIT 21 to just remove the issues with large numbers. I have found it quite helpful to have the exact number when it's small - especially when debugging issues with strange joins etc.

I think in most cases `20` will cover the number of returned objects (missing joins or sth) but in the same time it is not a big difference from a performance point of view. It can be helpful to get an exact number of returned objects, e.g. in Sentry reports.

**Comment by apollo13** [comment]
> but in the same time it is not a big difference from a performance point of view. 

Do we actually know that? That said, are we able to detect if .get tries to filter for a unique column? Ie the pk or so; then we wouldn't need the limit at all.

**Comment by shaib** [comment]
I believe "it is not a big difference from a performance point of view" mostly because it's an error case -- so, not supposed to be on any hot-spot. For error cases, I prefer to optimize for information over speed, as long as the costs are not excessive.

**Comment by dikamilo** [comment]
So `get_object` in `SingleObjectMixin` will be broken now since there is a case where `.get()` should return single object? And probably same with `get_object_or_404` ? Or I'm mistaken?

**Comment by felixxm** [comment]
@dikamilo `get()` returns a single objects and this behavior will not change. This patch is only an optimization in constructing a query for `get()`. Previously we retrieve all rows then we check number of rows and return single row or raise an exception. After this patch we will retrieve **limited** (e.g. **20**) rows instead of **all**.

**Comment by amir-hadi** [comment]
Will work on this during the weekend and update the PR. Also will mention the original authors using the Co-Authored-By header I guess? Is that correct @felixxm?

**Comment by felixxm** [comment]
@amir-hadi Sounds good, thanks :+1: 

@apollo13 Does 21 work for you? or maybe would you like to restart 6 years old discussion :wink:?

**Comment by apollo13** [comment]
> @apollo13 Does 21 work for you? or maybe would you like to restart 6 years old discussion wink? 

I do not have the performance metrics to make a good argument against them (time …). That said, I'd expect it to be slower with 21 than with 2 (especially if you use select_related), while the extra information of how many rows are returned is useless for me (at least I never needed it since I started working with django and actually assumed that in the error case [if one actually evaluated the error msg] we'd issue a count to get this information).

That said, 21 is better than fetching all of them, so I will not stand in the way of merging it. Personally I'd go with 2.

**Comment by felixxm** [comment]
@amir-hadi Do you have time to finish this patch? if not I can do this for you.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
I pushed minor edits, rebased from master and fixed issue on databases that don't support `SELECT FOR UPDATE` with limits (e.g. Oracle).

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by carltongibson** [comment]
Hi @amir-hadi. Thanks for the effort. Great stuff. Welcome aboard! 💃

**Comment by FrankApiyo** [comment]
I wish this `MAX_GET_RESULTS` was customizable with a config/setting

**Pull request merged**